rfcs/0001-authentication/text.md

```diff
@@ -0,0 +1,383 @@
+is more about where the authentication feature is initiated / implemented: The cellxgene server
+or the cellxgene client. I prefer the cellxgene server approach for the following reasons:
+
+1. In the client approach, more configuration parameters (client_id and domain) are needed by
+   the cellxgene client
+2. More specialized code is needed on the cellxgene client to start the authentication flow.
+   Keeping with the modular goals of cellxgene, this code would need to be implemented in a way
+   where it would not require auth0, but could be done in an authentication neutral way. This
+   could be implemented using the customization features in cellxgene (inline_scripts), or some
+   other way. The point being that there is a little more complexity with this approach, and I am
+   unaware of benefits over the server side approach.
+3. Keeping the authentication in the server could make it easier to support different types of
+   clients. Currently the only client is the cellxgene browser client. In the future there could
+   be an api client that accesses cellxgene programmatically through its REST api.
+
+### Cellxgene could use a popup approach that does not require a page navigation
+
+Auth0 provides a popup approach that avoids the page navigation to the Auth0 Universal Login
+page. This has the advantage that the client side state of cellxgene is not reset during the
+login process. This is very appealing since it makes the user experience nicer.
+
+However, the popup approach is intrinsically not as secure. It uses the Implicit Grant Flow,
+which is now longer recommended
+([https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-2.1.2](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-2.1.2)):
+“The implicit grant (response type "token") and other response types causing the authorization
+server to issue access tokens in the authorization response are vulnerable to access token
+leakage and access token replay as described in [Section 3.1](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-3.1), [Section 3.2](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-3.2), [Section 3.3](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-3.3), and [Section 3.6](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-3.6). In order to avoid these issues, Clients SHOULD NOT use
+the implicit grant and any other response type causing the authorization server to issue an
+access token in the authorization response.”
+
+The popup experience may also depend on the client’s browser and security settings, which could
+lead a larger testing and support burden.
+
+So, the consensus was to go with the “regular app” approach, and strongly encourage the user to
+login before doing work.
+
+## References
+
+[0][location to the authorization code flow diagram](https://auth0.com/docs/architecture-scenarios/web-app-sso/part-1)
```

Comment:
@bmccandless if we want the link text not be lowercase, we'll need to add `\` before the number like below, otherwise linter will force lowercase! Thank you!

```md
## References
\[0][IETF's Github Process](https://tools.ietf.org/html/draft-ietf-git-using-github-06)
\[1][Google Developer Documentation Style Guide](https://developers.google.com/style/highlights#introduction_1)
\[2][Data Coordination Platform RFC Process](https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0001-rfc-process.md)
```